ts/Studio.ts

```diff
@@ -1305,6 +1319,163 @@ export class Studio {
         }
     }
 
+    private bulkFrames(table: HTMLTableElement, framesButton: HTMLButtonElement) {
+        this.setTitle('Bulk Frames');
+
+        let thead = document.createElement('thead');
+        table.appendChild(thead);
+        let tr = document.createElement('tr');
+        thead.appendChild(tr);
+        for (const columnHeader of ['Action', 'Input']) {
+            const th = document.createElement('th');
+            th.textContent = columnHeader;
+            tr.appendChild(th);
+        }
+        const tbody = document.createElement('tbody');
+        table.appendChild(tbody);
+
+        // Move Frame
+        tr = document.createElement('tr');
+        tbody.appendChild(tr);
+        let td = document.createElement('td');
+        td.textContent = 'Move Frame';
+        tr.appendChild(td);
+        td = document.createElement('td');
+        let table2 = document.createElement('table');
+        table2.classList.add('table', 'table-borderless', 'mb-0');
+        td.appendChild(table2);
+        tr.appendChild(td);
+
+        // Move Frame Actions
+        const liftFrames = (value: number) => {
+            this.railroad.frames.forEach((frame: Frame) => {
+                frame.location.z += value;
+            });
+            framesButton.dispatchEvent(new Event('click'));
+            return value;
+        };
+
+        thead = document.createElement('thead');
+        table2.appendChild(thead);
+        tr = document.createElement('tr');
+        thead.appendChild(tr);
+        for (const columnHeader of ['Axis', 'Value']) {
+            const th = document.createElement('th');
+            th.textContent = columnHeader;
+            tr.appendChild(th);
+        }
+        let tbody2 = document.createElement('tbody');
+        table2.appendChild(tbody2);
+        tr = document.createElement('tr');
+        tbody2.appendChild(tr);
+        td = document.createElement('td');
+        td.textContent = 'Z';
+        tr.appendChild(td);
+        td = document.createElement('td');
+        tr.appendChild(td);
+        const options: InputTextOptions = {min: '0'};
+        td.appendChild(editNumber(this, 0, options, liftFrames));
+
+        // Refill Consumables
+        tr = document.createElement('tr');
+        tbody.appendChild(tr);
+        td = document.createElement('td');
+        td.textContent = 'Refill Consumables';
+        tr.appendChild(td);
+        td = document.createElement('td');
+        table2 = document.createElement('table');
+        table2.classList.add('table', 'table-borderless', 'mb-0');
+        td.appendChild(table2);
+        tr.appendChild(td);
+
+        // Refill Consumables Action
+        const refillFuel = (): void => {
+            this.railroad.frames.forEach((frame: Frame) => {
+                if (isFrameType(frame.type)) {
+                    const frameDef: FrameDefinition = frameDefinitions[frame.type];
```

Comment:
You can remove a layer of nesting by inverting this condition and taking an early exit:

```suggestion
                if (!isFrameType(frame.type)) return;
                const frameDef: FrameDefinition = frameDefinitions[frame.type];
                ...
```